### PR TITLE
Refactor connection specification and connection handler

### PR DESCRIPTION
**Pull request by casperisfine**
Followup: https://github.com/rails/rails/pull/37296

There is a lot of tests to fix on this PR, but I tested it against a small application, it's mostly functional. I'd just like to validate the direction it's taking before investing a lot more time into it.

### Current state

Right now, here's what a two DB setup, each with one writer and one reader looks like:

```ruby
class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true
  connects_to database: { writing: :primary, reading: :primary_replica }
end

class AnimalBase < ApplicationRecord
  self.abstract_class = true
  connects_to database: { writing: :animals, reading: :animals_replica }
end
```

With the following state:

```ruby
>> ApplicationRecord.connection_specification_name
=> "primary"
>> AnimalBase.connection_specification_name
=> "AnimalBase"
>> ActiveRecord::Base.connection_handlers.keys
=> [:writing, :reading]
>> ActiveRecord::Base.connection_handlers.fetch(:reading).connection_pool_names
=> ["primary", "AnimalBase"]
```

So in other words, there's one `ConnectionHandler` instance per `role`. And each `ConnectionHandler` connects to different "database schemas"

### Where this PR is trying to go

IMO it should be the inverse, there should be one `ConnectionHandler` per "database schema" (`primary` vs `animals`), and it should handle all the roles of that schema (`writing`, `reading`, etc).

With this PR:

```ruby
>> ApplicationRecord.connection_handler.connection_pool_names
=> [:writing, :reading]
>> AnimalBase.connection_handler.connection_pool_names
=> [:writing, :reading]
```

Whenever you call `connects_to` on a model, it assign a new `ConnectionHandler` on it, and from there the inheritance chain take care of accessing the right set of connections.

Also rather then switching handlers, we only switch the currently selected role:

```ruby
>> AnimalBase.current_role
=> :writing
>> AnimalBase.connected_to(role: :reading) { AnimalBase.current_role }
=> :reading
>> AnimalBase.connected_to(role: :reading) { ApplicationRecord.current_role }
=> :writing
```

@matthewd @eileencodes @seejohnrun at this stage I'd like your opinions here, because It's heavily changing how multidb connection works, and I think at least part of the backward compatibility will have to be broken (not 100% sure yet).

cc @rafaelfranca @Edouard-chin @etiennebarrie

### CODE REVIEW

**Comment by matthewd** [comment]
I agree it makes sense, within the context of our new object hierarchy, that there is a level of object that owns a particular DB schema (and later, in turn, a particular [shard] instance of that schema) -- having several unconnected objects trying to cross-manage that seems like a pain.

However, while I see advanced use-case value in being able to change the active role on a per-handler basis, I think it's important for the simple case to maintain the concept of a [thread-]global current role -- not necessarily actively switching between objects, but perhaps with https://github.com/rails/rails/pull/37388/files#diff-7735c67f539e7d1e2908a4d5d8909c24R1013 looking more like:
```ruby
current_roles.fetch(object_id) do
  # (note this is not an assignment, so it'll automatically follow any changes in the global setting)
  global = Thread.current.thread_variable_get(:ar_connection_role)
  role_to_config.key?(global) ? global : Base.writing_role
end
```

**Comment by eileencodes** [comment]
I think that there's a lot of value in doing it this new way, but I'm really worried about the big changes this will require to GitHub's codebase, especially after doing some pretty major changes just to get to 3-tier configs working.

The concern I have here is that we won't be able to maintain both versions in GitHub for 6.0 and 6.1 and that maintaining 2 ways to swap databases in our application will be a lot to ask of feature teams. I don't think we can outright break the API and expectations that were set in 6.0.

**Comment by casperisfine** [comment]
@eileencodes can I ask which APIs you are using?

`connected_to` obviously. But do you reach for `connection_specification_name` ? `Base.connection_handlers` ? Any others ?


@matthewd totally agree, but seems like you missed that `current_roles` is on a per thread registry: https://github.com/rails/rails/pull/37388/files#diff-7735c67f539e7d1e2908a4d5d8909c24R1147

**Comment by matthewd** [comment]
@casperisfine no I saw that `current_roles` is per-thread-per-handler; I'm suggesting that the fallback should be a global-per-thread value instead of being fixed to `writing`. So we can still globally switch to readonly mode (which I think is one the APIs that GitHub uses, and also the behaviour of the built-in middleware?) rather than having to do so on each handler separately.

**Comment by eileencodes** [comment]
The way I'm reading this is we'll need to change our application like the below to maintain both API's while since 6.0 and 6.1 will behave differently. This just isn't sustainable for feature teams to keep up with.

```
if GitHub.rails_6_0?
  ActiveRecord::Base.connected_to(role: x) do
    SomeModel.where(id: 1)
    SomeOtherModelNowThatWeCantUseHandlers.where(id: 2)
  end
else
  SomeModelName.connected_to(role: x) do
    SomeModel.where(id: 1)
  end

  SomeOtherModelNowThatWeCantUseHandlers.connected_to(role: x) do
    SomeOtherModelNowThatWeCantUseHandlers.where(id: 2)
  end
end
```

We also do some testing with `ActiveRecord::Base.connection_handlers` but that's not as a big deal as the above. We also do use `connection_specification_name` for migrations but I'm trying to get rid of that code in favor of the "Rails way". I'm mostly concerned that it's untenable for us to change how we use the `connected_to` API.

**Comment by eileencodes** [comment]
Before we knew all queries inside of `connected_to(role: :reading)` were going to be reading queries and we can then just let the models handle picking up their correct connection. This change would mean we need to explicitly call the `connected_to` API for each query we need.

**Comment by casperisfine** [comment]
@matthewd Oh! I think I get it now. And it also seems that it would indeed answer @eileencodes concerns.

@eileencodes just to make sure we're on the same page:

The current state of the PR doesn't require to use `connected_to` on a per-model basis, just on a per-db ones, so if you connect to two distinct DBs, two nested `connected_to` should be enough.

However, I think I now understand how you need the API to be used. So would you be ok with this:

```ruby
ActiveRecord::Base.connected_to(role: :reading) do
  ApplicationRecord.current_role # => :reading
  AnimalBase.current_role # => :reading
end

AnimalBase.connected_to(role: :reading) do
  ApplicationRecord.current_role # => :writing
  AnimalBase.current_role # => :reading
end
```

In short, `ActiveRecord::Base.connected_to` change the global (per thread actually) setting, but calling it on a subclass, only change it for that model handler.

**Comment by eileencodes** [comment]
The update you sent will work for us as we use `ActiveRecord::Base.connected_to` everywhere right now. I'm a little concerned that it's confusing for users, and maybe should log a warning if someone calls a model that inherits from `ApplicationRecord` from inside the context of `AnimalBase.connected_to` block?

```ruby
ActiveRecord::Base.connected_to(role: :reading) do
  ApplicationRecord.current_role # => :reading
  AnimalBase.current_role # => :reading
end

AnimalBase.connected_to(role: :reading) do
  ApplicationRecord.current_role # => :writing
  AnimalBase.current_role # => :reading
end
```

One question I have about the above is would exiting the block return the queries to default to Ar::Base? Or would exiting the block leave the requests on `AnimalBase.connected_to`?

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 1]
```diff
@@ -126,19 +126,14 @@ def migration_context # :nodoc:
...
                                   define_singleton_method(:name) { name }
                                   define_singleton_method(:to_s) { name }
 
-                                  connection_handler.connection_pool_names.each do |pool_name|
-                                    if conn.pool == connection_handler.retrieve_connection_pool(pool_name)
-                                      self.connection_specification_name = pool_name
-                                      break
-                                    end
-                                  end
+                                  establish_connection(db_config.configuration_hash)
```
Comment (seejohnrun) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]: We should be able to pass `db_config` here directly

File: activerecord/lib/active_record/connection_adapters/resolver.rb [hunk 2]
```diff
@@ -76,12 +81,12 @@ def spec(config)
       #   Resolver.new({}).resolve("postgresql://localhost/foo")
       #   # => DatabaseConfigurations::UrlConfig.new(config: {"adapter" => "postgresql", "host" => "localhost", "database" => "foo"})
       #
-      def resolve(config_or_env, pool_name = nil)
+      def resolve(config_or_env, _pool_name = nil)
```
Comment (seejohnrun) [inline activerecord/lib/active_record/connection_adapters/resolver.rb]: Will we be removing this entirely?

Comment (dylanahsmith) [reply activerecord/lib/active_record/connection_adapters/resolver.rb]: Seems like we should at least deprecate the `_pool_name` parameter if we can't remove it yet for backwards compatibility reasons.

Comment (eileencodes) [reply activerecord/lib/active_record/connection_adapters/resolver.rb]: @dylanahsmith `resolve` is a private API, we don't need to deprecate it.

Comment (dylanahsmith) [reply activerecord/lib/active_record/connection_adapters/resolver.rb]: Great.  We should just remove it in that case.

```suggestion
      def resolve(config_or_env)
```

File: activerecord/lib/active_record/connection_adapters/resolver.rb [hunk 3]
```diff
@@ -53,6 +43,21 @@ def spec(config)
...
+      # == Example
+      #
+      #   config = { "production" => { "host" => "localhost", "database" => "foo", "adapter" => "sqlite3" } }
+      #   spec = Resolver.new(config).spec(:production)
+      #   spec.db_config.configuration_hash
+      #   # => { host: "localhost", database: "foo", adapter: "sqlite3" }
+      #
+      def spec(config)
+        db_config = lookup(config)
         ConnectionSpecification.new(db_config.configuration_hash.delete(:name) || "primary", db_config)
```
Comment (dylanahsmith) [inline activerecord/lib/active_record/connection_adapters/resolver.rb]: Since the `resolve` call in `lookup` is no longer setting the `:name` key, I don't think it will even be there anymore.  If we were going to keep this method, then I think this line needs to be updated to just get the name from the `db_config` directly.  However, it looks like this method is considered private and this PR makes it unused, so we could just remove it entirely.

**Comment by casperisfine** [comment]
> maybe should log a warning if someone calls a model that inherits from `ApplicationRecord` from inside the context of `AnimalBase.connected_to` block?

Well, the way I see it, you should be able to call `connected_to` very high up the stack (like a `middleware` or `around_action`), so IMO such warning would catch lots of false positives. 

IMO there is a use case for having read only access to one db, but write access to the other. 

> One question I have about the above is would exiting the block return the queries to default to Ar::Base? Or would exiting the block leave the requests on AnimalBase.connected_to?

Not sure I understand your question. When you exit the block you are back to the previous state, so for each model you are back to query the default role (most likely `writing`).

File: activerecord/lib/active_record/connection_adapters/resolver.rb (hunk 2 above)
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/resolver.rb]: I'd like to, but I'm not sure it's an API breakage or not, so I left it.

Comment (rafaelfranca) [reply activerecord/lib/active_record/connection_adapters/resolver.rb]: 🔥 it. Anything with `:nodoc` or inside a `nodoc` constant can be removed without deprecation.

**Comment by casperisfine** [comment]
I just pushed a quick change such as:

```ruby
>> ActiveRecord::Base.connected_to(role: :reading) { p [ActiveRecord::Base.current_role, ApplicationRecord.current_role, AnimalBase.current_role] }
[:writing, :writing, :writing]
=> [:writing, :writing, :writing]
>> ApplicationRecord.connected_to(role: :reading) { p [ActiveRecord::Base.current_role, ApplicationRecord.current_role, AnimalBase.current_role] }
[:writing, :reading, :writing]
=> [:writing, :reading, :writing]
>> AnimalBase.connected_to(role: :reading) { p [ActiveRecord::Base.current_role, ApplicationRecord.current_role, AnimalBase.current_role] }
[:writing, :writing, :reading]
=> [:writing, :writing, :reading]
```

I'll handle other suggestions and start fixing tests.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 4]
```diff
@@ -125,24 +125,31 @@ def migration_context # :nodoc:
...
       end
 
+      def find_connection_handler
+        ActiveRecord::Base.connection_handlers.each do |name, handler|
+          handler.connection_pools.each do |pool|
+            return name.constantize if pool == self.pool
+          end
+        end
+        raise ConnectionNotEstablished, "Couldn't find an active connection for #{pool.db_config.spec_name.inspect}"
+      end
```
Comment (casperisfine) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]: This change might seem a bit weird, and I agree, however I'm afraid it's a lesser evil until we get further.

Later on we should have an object tree that represent the different databases, so it will be easy to simply create a sublcass, and give it a reference to the schema it maps to.

However right now, the topology is still very much held by the model classes, so only proper way to access the `animals` connection, is to inherit from `AnimalBase`, hence this change. The `constantize` is just an unfortunate consequence of this.

But again, I'm confident we'll be able to clean that up down the road, so I hope it's not too much of a deal breaker.

File: activerecord/test/cases/test_case.rb [hunk 5]
```diff
@@ -30,6 +30,25 @@ def teardown
       SQLCounter.clear_log
     end
 
+    def before_setup
```
Comment (Edouard-chin) [inline activerecord/test/cases/test_case.rb]: The change we did in this PR revealed a bunch of leaks in the ActiveRecord test suite. One of the main issue
was tests doing something like

```ruby
test 'foo' do
  ActiveRecord::Base.connects_to(writing: :arunit2, reading: :other_db)
end
```

The problem with this is that after that test runs, `ActiveRecord::Base` connection_handler will hold the connection on the activerecord_unittest2 db (and not the regular activerecord_unittest).

I started fixing tests independently but ended up in a huge rabbit hole so I opted for a "global" state reset instead of doing a `@previous_connection_handlers = ..., ensure ...` in a **lot** of places

**Comment by Edouard-chin** [comment]
Hey friends  , this PR is ready to be reviewed, we managed to get a green build 🙇

**Review by eileencodes — COMMENTED** [review]
Thanks for working on this 😄 I did a quick review, but want to look closer when I'm back in the office on Friday.

We definitely want to move away from `connection_specification_name` because it ties a connection config to a class. I think by moving that to the handler key we're not fixing the underlying problem which is a config has no idea what class it belongs to until it gets connected. I'm worried this implementation hides that "problem" and makes it harder to drop the class entirely.

There's a small bug in here as well. If we continue in this direction we need to treat "ActiveRecord::Base" and "ApplicationRecord" the same otherwise we end up with duplicate connections. The problem with that can be seen here:

On boot in development mode:

```
>> ActiveRecord::Base.connection_handlers
=> {"ActiveRecord::Base"=>#<ActiveRecord::ConnectionAdapters::ConnectionHandler:0x00007fbc767c4f18 @roles=#<Concurrent::Map:0x00007fbc767c4e78 entries=1 default_proc=nil>>}
```

After calling another connection class:

```
>> MealBase
=> MealBase(abstract)
>> ActiveRecord::Base.connection_handlers
=> {"ActiveRecord::Base"=>#<ActiveRecord::ConnectionAdapters::ConnectionHandler:0x00007fbc767c4f18 @roles=#<Concurrent::Map:0x00007fbc767c4e78 entries=1 default_proc=nil>>, "ApplicationRecord"=>#<ActiveRecord::ConnectionAdapters::ConnectionHandler:0x00007fbc7c025c20 @roles=#<Concurrent::Map:0x00007fbc7c025bd0 entries=2 default_proc=nil>>, "MealBase"=>#<ActiveRecord::ConnectionAdapters::ConnectionHandler:0x00007fbc78427598 @roles=#<Concurrent::Map:0x00007fbc78427548 entries=2 default_proc=nil>>}
```

My biggest concern is the ways in which this is changing the public API and how that will affect applications and libraries

File: activerecord/lib/active_record/connection_adapters/role.rb [hunk 6]
```diff
@@ -5,7 +5,7 @@ module ConnectionAdapters
     class Role # :nodoc:
       include Mutex_m
 
-      attr_reader :db_config, :connection_specification_name
+      attr_reader :db_config, :role
```
[the remark this answers is not in the document]
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/role.rb]: Agreed, that's an oversight, we actually mentioned calling it `:name` with @Edouard-chin, but forgot. I'll fix.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 7]
```diff
@@ -125,24 +125,31 @@ def migration_context # :nodoc:
...
-                                  define_singleton_method(:to_s) { name }
-
-                                  connection_handler.connection_pool_names.each do |pool_name|
-                                    if conn.pool == connection_handler.retrieve_connection_pool(pool_name)
-                                      self.connection_specification_name = pool_name
-                                      break
-                                    end
+                                superclass = find_connection_handler
+                                if superclass == ActiveRecord::Base
+                                  ActiveRecord::SchemaMigration
```
[the remark this answers is not in the document]
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]: It doesn't *have* to, but it makes it so that `ActiveRecord::Base.connection.schema_migration == ActiveRecord::SchemaMigration`, which is nicer than getting an anonymous class, and better for backward compat.

All this is very temporary though, I really don't like the current `schema_migration` implementation, it's just hard to do better for now.

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 8]
```diff
@@ -1097,15 +1103,15 @@ def flush_idle_connections!
...
-      def retrieve_connection(spec_name) #:nodoc:
-        pool = retrieve_connection_pool(spec_name)
+      def retrieve_connection(role) #:nodoc:
+        pool = retrieve_connection_pool(role)
 
         unless pool
           # multiple database application
-          if ActiveRecord::Base.connection_handler != ActiveRecord::Base.default_connection_handler
-            raise ConnectionNotEstablished, "No connection pool with '#{spec_name}' found for the '#{ActiveRecord::Base.current_role}' role."
+          if connection = ActiveRecord::Base.connection_handler.connection_pool_list.first
```
[the remark this answers is not in the document]
Comment (Edouard-chin) [reply activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: Thanks for pointing that out, just realised it it wasn't 100% correct.

> If there's a pool in the list and we have a connection then why does this end up raising?

It end up raising because there is no pool for the given role. What I want to do here is grab any connection pool (doesn't matter which role) in order to get the database name. This comes with the assumptions that each roles (writing, reading) on the handler will connect to the same db

File: activerecord/lib/active_record/connection_handling.rb [hunk 9]
```diff
@@ -192,14 +185,13 @@ def connection
       retrieve_connection
     end
 
-    attr_writer :connection_specification_name
+    attr_writer :connection_handler
```
Comment (dylanahsmith) [inline activerecord/lib/active_record/connection_handling.rb]: Can we instead just have the model specify the schema name?

```ruby
  class_attribute :schema_name

  alias_method :connection_specification_name, :schema_name
  alias_method :connection_specification_name=, :schema_name=
  deprecate :connection_specification_name, connection_specification_name=
```

then detach the connection handling from the model and just look it up from the model using the `schema_name`.  That way we can do things like switch the connection for a schema as part of shard migrations and the connection switch affect the sharded models.

File: activerecord/lib/active_record/connection_adapters/role.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/role.rb]: We  fixed that already.
